File: ultimate/MessageServiceFacade.java
package ultimate;

public class MessageServiceFacade {

    private MessageHandlerCachingProxy messageHandlerCachingProxy;
    private MessageRepository messageRepository;
    private MessageAnalyzeService messageAnalyzeService;
    public MessageServiceFacade(MessageHandlerCachingProxy messageHandlerCachingProxy,
                                MessageRepository messageRepository, MessageAnalyzeService messageAnalyzeService) {
        this.messageHandlerCachingProxy = messageHandlerCachingProxy;
        this.messageRepository = messageRepository;
        this.messageAnalyzeService = messageAnalyzeService;
    }

    public String handle(String messageToHandle) {
        messageAnalyzeService.analyzeUnFormattedMessage(messageToHandle);
        Logger.getInstance().logInfo("Fetching the message formatter...");
        MessageFormatter messageFormatter = MessageFormatterFactory.getMessageFormatterInstance(messageToHandle);
        String formattedMessage = messageFormatter.formatMessage(messageToHandle);
        Logger.getInstance().logInfo("Message is formatted successfully with: " + messageFormatter.getType());
        messageAnalyzeService.analyzeFormattedMessage(messageToHandle);
        messageRepository.save(formattedMessage);
        return messageHandlerCachingProxy.processMessage(formattedMessage);
    }
}

File: typescript/patterns/ultimate-design-pattern-challenge/message-service-facade.ts
class MessageServiceFacade {
  private messageHandlerCachingProxy: MessageHandlerCachingProxy;
  private messageRepository: MessageRepository;
  private messageAnalyzeService: MessageAnalyzeService;

  constructor(
    messageAnalyzeService: MessageAnalyzeService,
    messageRepository: MessageRepository,
    messageHandlerCachingProxy: MessageHandlerCachingProxy,
  ) {
    this.messageAnalyzeService = messageAnalyzeService;
    this.messageRepository = messageRepository;
    this.messageHandlerCachingProxy = messageHandlerCachingProxy;
  }

  handle(messageToHandle: string): string {
    Logger.getInstance().logInfo("Analyzing un-formatted message...");
    this.messageAnalyzeService.analyzeUnFormattedMessage(messageToHandle);

    Logger.getInstance().logInfo("Fetching the message formatter...");
    const messageFormatter =
      MessageFormatterFactory.getMessageFormatterInstance(messageToHandle);

    Logger.getInstance().logInfo(
      `Message is formatted successfully with: ${messageFormatter.getType()}`,
    );
    const formattedMessage = messageFormatter.formatMessage(messageToHandle);

    Logger.getInstance().logInfo("Analyzing formatted message...");
    this.messageAnalyzeService.analyzeFormattedMessage(formattedMessage);

    Logger.getInstance().logInfo("Saving formatted message into database...");
    this.messageRepository.save(formattedMessage);

    Logger.getInstance().logInfo("Processing message through caching...");
    return this.messageHandlerCachingProxy.processMessage(formattedMessage);
  }
}
